### PR TITLE
#223 Avoid NullPointerException while parsing response

### PR DESCRIPTION
**Pull request by Nowheresly**
- check for null are added before using data field from javax.xml.stream.events.Character
see issue #223

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@Nowheresly__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        This seems like a small (but important) contribution, so no contribution license agreement is required at this point. Real humans will now evaluate your PR.
    </span>

TTYL, AZPRBOT;

**Comment by Nowheresly** [comment]
I added the unit test as a separate commit since the unit test requires a new dependency, namely mockito.
This dependency has a test scope and thus is only required for test purposes.

File: src/test/java/microsoft/exchange/webservices/data/EwsXmlReaderTest.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
+package microsoft.exchange.webservices.data;
```
Comment (serious6) [inline src/test/java/microsoft/exchange/webservices/data/EwsXmlReaderTest.java]: please add a license-header like shown [here](https://github.com/OfficeDev/ews-java-api/blob/master/src/main/java/microsoft/exchange/webservices/data/AbsoluteDateTransition.java). Plz remind the line between it and the package.

File: src/test/java/microsoft/exchange/webservices/data/EwsXmlReaderTest.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    impl = new EwsXmlReader(new ByteArrayInputStream(("<?xml version=\"1.0\" encoding=\"UTF-8\"?>"
+        + "<test></test>").getBytes("UTF-8")));
+    MockitoAnnotations.initMocks(this);
+
+  }
+
+  @Test
+  public void testReadValueWhenCharacterDataIsNull() throws Exception {
+
+    doReturn(false).when(presentEvent).isStartElement();
```
Comment (serious6) [inline src/test/java/microsoft/exchange/webservices/data/EwsXmlReaderTest.java]: should we maybe split the mocking code from the underlying test?

**Comment by serious6** [comment]
hi @Nowheresly thanks for you contribution, I think mockito makes sense. I have also seen some alternatives on the web. Maybe you can have some words about why mockito is the right one to choose?
Please also take into account that we should have some tests (in the future) which will enable us to make some unit-tests on other api-procedures.

File: src/test/java/microsoft/exchange/webservices/data/EwsXmlReaderTest.java (hunk 1 above)
Comment (Nowheresly) [inline src/test/java/microsoft/exchange/webservices/data/EwsXmlReaderTest.java]: Done. Sorry for the miss...

File: src/test/java/microsoft/exchange/webservices/data/EwsXmlReaderTest.java (hunk 2 above)
Comment (Nowheresly) [inline src/test/java/microsoft/exchange/webservices/data/EwsXmlReaderTest.java]: I usually write unit tests like this one:
- as instance variables, I declare all mocks needed for the test. The instance variable called 'impl' is *always* an instance of the class under test. All these instances variables *must* be refreshed before each test method execution.
- I declare a setUp method annotated with '@Before' which is in charge of refreshing all instances variables before each test and that contains as well all generic mock configuration (ie: all mock configuration that can be shared with all test methods)
- finally, I declare test methods where I follow the convention //given //when //then (see http://en.wikipedia.org/wiki/Behavior_Driven_Development).

In the case of this specific test, I expect all my mock declarations to be useful and needed to write other test methods. That's why it makes sens to me to keep this mocking code inside the test. 
But I'm obviously open to other suggestion!

**Comment by Nowheresly** [comment]
Hi @serious6 
sure, there's a lot of alternatives to mockito: easymock, spock to name a few.
- spock (http://docs.spockframework.org/en/latest/#) is really really powerful. But since it's groovy based (a jvm scripting language) and the project currently does not used groovy at all...
- mockito is younger than its counterpart easymock. It is well designed and IMHO is simpler to use than easymock. You can see a comparison [here](https://code.google.com/p/mockito/wiki/MockitoVSEasyMock)

Finally, Mockito has some limitations (see the [FAQ](https://github.com/mockito/mockito/wiki/FAQ)) but has proven to be very useful.

**Comment by serious6** [comment]
@Nowheresly maybe you can squash some of the commits. This would provide a better overview.

**Comment by Nowheresly** [comment]
makes sense indeed !
you can find the new pull request here #231

**Comment by serious6** [comment]
you can do a squash as simple as:
* git reset --soft HEAD~4
* git commit (with the files you want to commit as one or more commits)
* git push --force

**Pull request closed without merging**